### PR TITLE
Retain list type when assigning to offset 1 of non-empty-list

### PR DESCRIPTION
**Pull request by herndlm**
Fixes https://github.com/phpstan/phpstan/issues/12131

I don't think we can detect this edge case inside `Type::setOffsetValueType()` because we need info of 2 different accessory types, right? Is `IntersectionType` the best place then I suppose?

### CODE REVIEW

File: tests/PHPStan/Rules/Properties/data/bug-12131.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	public function setAtZero(): void
+	{
+		$this->array[0] = 1;
+	}
+
+	public function setAtOne(): void
+	{
+		$this->array[1] = 1;
+	}
+}
```
Comment (staabm) [inline tests/PHPStan/Rules/Properties/data/bug-12131.php]: IMO this test should have a 

```
	public function setAtTwo(): void
	{
		$this->array[2] = 2;
	}
```

method and define expectations for it

Comment (herndlm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: you mean you want to see it (still) keep failing here? is that enough? but I can also add some type assertions and load it in NodeScopeResolverTest I guess. just not sure if it's really needed

Comment (herndlm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: or I add more specific list-related test cases to the generic test file for this rule: `tests/PHPStan/Rules/Properties/data/properties-assigned-types.php`

Comment (staabm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: > you mean you want to see it (still) keep failing here? is that enough?

yes.

> but I can also add some type assertions and load it in NodeScopeResolverTest I guess

sounds good to me

Comment (staabm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: > If I have a class with an array property declared as `list<T>` or `non-empty-list<T>` and if I try to assign a value to an existing element with a non-zero index in that list, e.g. `$array[1] = 1`, then I get an assign.propertyType error:
> `non-empty-array<int<0, max>, int> might not be a list.`

reading this from the error report, I also wonder whether its correct just to add hardcoded support for `1`.

Comment (staabm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: e.g. https://phpstan.org/r/b40e3747-909e-47bb-96df-2a69df6780fb

Comment (herndlm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: added more tests :)

> e.g. https://phpstan.org/r/b40e3747-909e-47bb-96df-2a69df6780fb

this is weird and seems to be something else. pretty sure the constantarraytype(builder) is taking care of this list properly.. https://phpstan.org/r/5364810d-df16-49e7-a465-ca63f0033f17 strange

Comment (herndlm) [reply tests/PHPStan/Rules/Properties/data/bug-12131.php]: ah, I see the `MutatingScope` is asked for the type of `$this->array` with a `OriginalPropertyTypeExpr` and that returns not the constant array but, surprise, the original property type :/ the scope would even know what `$this->array` is here, but I guess there is a good reason for doing it like it is done right now hmm.

UPDATE: only a couple of tests start failing if that behaviour is changed, but those cases become slightly weird. E.g. invalid assignments cause still an error, but then that invalid assignment leads to a prop type change which influences future assignments in the scope and either cause false positives/negatives or just different, sometimes weird types in error messages. it might be possible to improve the current behaviour though, maybe I'll keep looking a bit more..

**Comment by ondrejmirtes** [comment]
Thank you.

File: tests/PHPStan/Analyser/nsrt/list-type.php [hunk 2]
```diff
@@ -106,4 +106,29 @@ public function testUnset(array $list): void
...
+		assertType('list<int>', $list);
+		$list[0] = 17;
+		assertType('non-empty-list<int>&hasOffsetValue(0, 17)', $list);
+		$list[1] = 19;
+		assertType('non-empty-list<int>&hasOffsetValue(0, 17)&hasOffsetValue(1, 19)', $list);
+		$list[0] = 21;
+		assertType('non-empty-list<int>&hasOffsetValue(0, 21)&hasOffsetValue(1, 19)', $list);
+
+		$list[2] = 23;
+		assertType('non-empty-array<int<0, max>, int>&hasOffsetValue(0, 21)&hasOffsetValue(1, 19)&hasOffsetValue(2, 23)', $list);
```
Comment (herndlm) [inline tests/PHPStan/Analyser/nsrt/list-type.php:121]: wanted to delete this. because that other method below exists 🤦

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/nsrt/list-type.php:121]: There's always the next PR :)
